ouroboros-consensus/src/Ouroboros/Consensus/Node.hs

```diff
@@ -317,7 +321,7 @@ forkBlockProduction IS{..} =
             -- valid with respect to the current ledger state of the
             -- 'ChainDB'. Refer to the 'getTxs' documentation for more
             -- information.
-            _invalidTxs         <- syncState mempool
+            _                   <- pure $ syncState mempool
```

Comment:
This does nothing. It is a no-op.

The `syncState` is in `m` not `STM m`, which is why it didn't type without the `pure`. But with the `pure` we can prove via the monad laws that this has no effect! :-)

`forall x.  return x >>= \_ -> action = action`

This will need to be fixed if it hasn't been already. I think what we'll need to do is to finish the `atomically` after determining if we're the leader. Then when we are, we call `syncState` which we change to return the resulting mempool snapshot. We use that to compute the new block and call `ChainDB.addBlock`.

There's still a race between us getting the snapshot and adding the block, but that's ok. The `addBlock` deals with that race condition. But it does mean that it's a legitimate result for `addBlock` to end up not adding the block, if something sneaked in ahead. We should make sure the trace results for this use of `addBlock` reflect that. cc @mrBliss

Reply:
See #777 for a possible solution that still does everything in one transaction.

Maybe #777 is overkill, since we still have the race condition between producing a block and adding it.